### ISSUE TITLE
EML parsing error .. ?

### ISSUE THREAD

**Issue by silenius** [gbif/ipt#1356]
Hello,

I don't know if it's an $user mistake or an IPT bug, but it seems that one of the generated EML (XML) file is not well formed, this is the problematic dataset: http://ipt.biodiversity.aq/resource?r=cyano_16s

I took a look at the XML file and it seems not valid:

```
jcigar@orval:/usr/jails/java1/filer/ipt/ipt.biodiversity.aq/private/resources/cyano_16s/ > xmllint eml-3.0.xml 
eml-3.0.xml:139: parser error : PCDATA invalid Char value 21
ntaining the freshwater cyanobacterium Aphanizomenon flos-aquae. Toxicon 46, 555
                                                                               ^
eml-3.0.xml:151: parser error : PCDATA invalid Char value 14
(ASPA) No. 126 of Byers Peninsula, Livingston Island (62 340 3500 to 62 400 35
                                                        ^
eml-3.0.xml:151: parser error : PCDATA invalid Char value 14
(ASPA) No. 126 of Byers Peninsula, Livingston Island (62 340 3500 to 62 400 35
                                                                        ^
eml-3.0.xml:151: parser error : PCDATA invalid Char value 14
 126 of Byers Peninsula, Livingston Island (62 340 3500 to 62 400 3500 Sand 60
                                                                               ^
eml-3.0.xml:151: parser error : PCDATA invalid Char value 14
ninsula, Livingston Island (62 340 3500 to 62 400 3500 Sand 60 540 1400 to 61
                                                                               ^
eml-3.0.xml:152: parser error : PCDATA invalid Char value 14
sterile spatula, sealed in sterile plastic bags or tubes and stored frozen (-20 
                                                                               ^
```

Those are the problematic lines:

__line 139__:

```
jcigar@orval:/usr/jails/java1/filer/ipt/ipt.biodiversity.aq/private/resources/cyano_16s/ > awk 'FNR==139' eml-3.0.xml      
Saker, M. L., Jungblut, A. D., Neilan, B. A., Rawn, D. F. K. &amp; Vasconcelos, V. M. Detection of microcystin synthetase genes in health food supplements containing the freshwater cyanobacterium Aphanizomenon flos-aquae. Toxicon 46, 555562 (2005).</para>

jcigar@orval:/usr/jails/java1/filer/ipt/ipt.biodiversity.aq/private/resources/cyano_16s/ > awk 'FNR==139' eml-3.0.xml|od -c
0000000    S   a   k   e   r   ,       M   .       L   .   ,       J   u
0000020    n   g   b   l   u   t   ,       A   .       D   .   ,       N
0000040    e   i   l   a   n   ,       B   .       A   .   ,       R   a
0000060    w   n   ,       D   .       F   .       K   .       &   a   m
0000100    p   ;       V   a   s   c   o   n   c   e   l   o   s   ,    
0000120    V   .       M   .       D   e   t   e   c   t   i   o   n    
0000140    o   f       m   i   c   r   o   c   y   s   t   i   n       s
0000160    y   n   t   h   e   t   a   s   e       g   e   n   e   s    
0000200    i   n       h   e   a   l   t   h       f   o   o   d       s
0000220    u   p   p   l   e   m   e   n   t   s       c   o   n   t   a
0000240    i   n   i   n   g       t   h   e       f   r   e   s   h   w
0000260    a   t   e   r       c   y   a   n   o   b   a   c   t   e   r
0000300    i   u   m       A   p   h   a   n   i   z   o   m   e   n   o
0000320    n       f   l   o   s   -   a   q   u   a   e   .       T   o
0000340    x   i   c   o   n       4   6   ,       5   5   5 025   5   6
0000360    2       (   2   0   0   5   )   .   <   /   p   a   r   a   >
0000400   \n                                                            
0000401
```
__line 151-152__:
```

jcigar@orval:/usr/jails/java1/filer/ipt/ipt.biodiversity.aq/private/resources/cyano_16s/ > awk 'FNR>=151 && FNR<=152' eml-3.0.xml      
(ASPA) No. 126 of Byers Peninsula, Livingston Island (62 340 3500 to 62 400 3500 Sand 60 540 1400 to 61 130 0700 W), South Shetland Islands, Antarctic Peninsula, during an expedition in February 2009. 
Microbial mats were probed using a sterile spatula, sealed in sterile plastic bags or tubes and stored frozen (-20 C) for DNA extraction.</para>

jcigar@orval:/usr/jails/java1/filer/ipt/ipt.biodiversity.aq/private/resources/cyano_16s/ > awk 'FNR>=151 && FNR<=152' eml-3.0.xml|od -c
0000000    (   A   S   P   A   )       N   o   .       1   2   6       o
0000020    f       B   y   e   r   s       P   e   n   i   n   s   u   l
0000040    a   ,       L   i   v   i   n   g   s   t   o   n       I   s
0000060    l   a   n   d       (   6   2 016       3   4   0       3   5
0000100    0   0       t   o       6   2 016       4   0   0       3   5
0000120    0   0       S   a   n   d       6   0 016       5   4   0    
0000140    1   4   0   0       t   o       6   1 016       1   3   0    
0000160    0   7   0   0       W   )   ,       S   o   u   t   h       S
0000200    h   e   t   l   a   n   d       I   s   l   a   n   d   s   ,
0000220        A   n   t   a   r   c   t   i   c       P   e   n   i   n
0000240    s   u   l   a   ,       d   u   r   i   n   g       a   n    
0000260    e   x   p   e   d   i   t   i   o   n       i   n       F   e
0000300    b   r   u   a   r   y       2   0   0   9   .      \r  \n   M
0000320    i   c   r   o   b   i   a   l       m   a   t   s       w   e
0000340    r   e       p   r   o   b   e   d       u   s   i   n   g    
0000360    a       s   t   e   r   i   l   e       s   p   a   t   u   l
0000400    a   ,       s   e   a   l   e   d       i   n       s   t   e
0000420    r   i   l   e       p   l   a   s   t   i   c       b   a   g
0000440    s       o   r       t   u   b   e   s       a   n   d       s
0000460    t   o   r   e   d       f   r   o   z   e   n       (   -   2
0000500    0     016   C   )       f   o   r       D   N   A       e   x
0000520    t   r   a   c   t   i   o   n   .   <   /   p   a   r   a   >
0000540   \n                                                            
0000541
```

it seems that some escape characters are not properly escaped ? (025 in the first paste, 016 in the second)

**Comment by silenius** [comment]
Here is the stacktrace:

```
ERROR Failed to check if last published version of resource has been assigned a GBIF-supported license: An invalid XML character (Unicode: 0x15) was found in the element content of the document.
org.xml.sax.SAXParseException; lineNumber: 139; columnNumber: 238; An invalid XML character (Unicode: 0x15) was found in the element content of the document.
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1239)
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:643)
	at org.apache.commons.digester3.Digester.parse(Digester.java:1642)
	at org.apache.commons.digester3.Digester.parse(Digester.java:1701)
	at org.gbif.metadata.eml.EmlFactory.build(EmlFactory.java:120)
	at org.gbif.ipt.action.manage.OverviewAction.getLastPublishedVersionAssignedLicense(OverviewAction.java:1366)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at freemarker.ext.beans.BeansWrapper.invokeMethod(BeansWrapper.java:1458)
	at freemarker.ext.beans.SimpleMethodModel.exec(SimpleMethodModel.java:69)
	at freemarker.core.MethodCall._eval(MethodCall.java:65)
	at freemarker.core.Expression.eval(Expression.java:81)
	at freemarker.core.DefaultToExpression._eval(DefaultToExpression.java:84)
	at freemarker.core.Expression.eval(Expression.java:81)
	at freemarker.core.Environment.setMacroContextLocalsFromArguments(Environment.java:784)
	at freemarker.core.Environment.invoke(Environment.java:702)
	at freemarker.core.UnifiedCall.accept(UnifiedCall.java:83)
	at freemarker.core.Environment.visit(Environment.java:327)
	at freemarker.core.Environment.visit(Environment.java:333)
	at freemarker.core.Environment.visit(Environment.java:333)
	at freemarker.core.Environment.visit(Environment.java:333)
	at freemarker.core.Environment.visit(Environment.java:333)
	at freemarker.core.Environment.process(Environment.java:306)
	at freemarker.template.Template.process(Template.java:386)
	at org.apache.struts2.views.freemarker.FreemarkerResult.doExecute(FreemarkerResult.java:223)
	at org.apache.struts2.dispatcher.StrutsResultSupport.execute(StrutsResultSupport.java:191)
	at com.opensymphony.xwork2.DefaultActionInvocation.executeResult(DefaultActionInvocation.java:372)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:276)
	at com.opensymphony.xwork2.interceptor.DefaultWorkflowInterceptor.doIntercept(DefaultWorkflowInterceptor.java:168)
	at com.opensymphony.xwork2.interceptor.MethodFilterInterceptor.intercept(MethodFilterInterceptor.java:98)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at com.opensymphony.xwork2.validator.ValidationInterceptor.doIntercept(ValidationInterceptor.java:265)
	at org.apache.struts2.interceptor.validation.AnnotationValidationInterceptor.doIntercept(AnnotationValidationInterceptor.java:76)
	at com.opensymphony.xwork2.interceptor.MethodFilterInterceptor.intercept(MethodFilterInterceptor.java:98)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.apache.struts2.interceptor.DeprecationInterceptor.intercept(DeprecationInterceptor.java:41)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at com.opensymphony.xwork2.interceptor.ConversionErrorInterceptor.intercept(ConversionErrorInterceptor.java:138)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at com.opensymphony.xwork2.interceptor.ParametersInterceptor.doIntercept(ParametersInterceptor.java:229)
	at com.opensymphony.xwork2.interceptor.MethodFilterInterceptor.intercept(MethodFilterInterceptor.java:98)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at com.opensymphony.xwork2.interceptor.ParametersInterceptor.doIntercept(ParametersInterceptor.java:229)
	at com.opensymphony.xwork2.interceptor.MethodFilterInterceptor.intercept(MethodFilterInterceptor.java:98)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.apache.struts2.interceptor.MultiselectInterceptor.intercept(MultiselectInterceptor.java:73)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.apache.struts2.interceptor.DateTextFieldInterceptor.intercept(DateTextFieldInterceptor.java:125)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.apache.struts2.interceptor.CheckboxInterceptor.intercept(CheckboxInterceptor.java:91)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at com.opensymphony.xwork2.interceptor.PrepareInterceptor.doIntercept(PrepareInterceptor.java:171)
	at com.opensymphony.xwork2.interceptor.MethodFilterInterceptor.intercept(MethodFilterInterceptor.java:98)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.apache.struts2.interceptor.ServletConfigInterceptor.intercept(ServletConfigInterceptor.java:164)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at com.opensymphony.xwork2.interceptor.ExceptionMappingInterceptor.intercept(ExceptionMappingInterceptor.java:189)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at com.opensymphony.xwork2.interceptor.I18nInterceptor.intercept(I18nInterceptor.java:140)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.gbif.ipt.struts2.RedirectMessageInterceptor.doIntercept(RedirectMessageInterceptor.java:120)
	at com.opensymphony.xwork2.interceptor.MethodFilterInterceptor.intercept(MethodFilterInterceptor.java:98)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.gbif.ipt.struts2.RequireManagerInterceptor.intercept(RequireManagerInterceptor.java:91)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.gbif.ipt.struts2.SetupAndCancelInterceptor.intercept(SetupAndCancelInterceptor.java:84)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.gbif.ipt.struts2.ResourceSessionInterceptor.intercept(ResourceSessionInterceptor.java:41)
	at com.google.inject.struts2.Struts2Factory$ProvidedInterceptor.intercept(Struts2Factory.java:215)
	at com.opensymphony.xwork2.DefaultActionInvocation.invoke(DefaultActionInvocation.java:245)
	at org.apache.struts2.impl.StrutsActionProxy.execute(StrutsActionProxy.java:54)
	at org.apache.struts2.dispatcher.Dispatcher.serviceAction(Dispatcher.java:575)
	at org.apache.struts2.dispatcher.ng.ExecuteOperations.executeAction(ExecuteOperations.java:81)
	at org.apache.struts2.dispatcher.ng.filter.StrutsPrepareAndExecuteFilter.doFilter(StrutsPrepareAndExecuteFilter.java:99)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
	at org.gbif.ipt.struts2.ResponseHeaderFilter.doFilter(ResponseHeaderFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:66)
	at org.gbif.ws.server.filter.XSSFilter.doFilter(XSSFilter.java:48)
	at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
	at org.gbif.ipt.struts2.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:43)
	at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
	at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:118)
	at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:113)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:94)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:504)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:502)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1132)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:684)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1533)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1489)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
ERROR EML file version #3.0 for resource cyano_16s couldn't be parsed
```

**Comment by silenius** [comment]
it seems to work after (re-)editing the resource.. maybe it was the remnants of an old bug?

**Comment by kbraak** [comment]
Thanks @silenius 

When you re-edited the resource, did you remove all the offending escape characters? 

I am glad that when you published a new version (3.1) of this resource the IPT produced a well formed EML document.

The previous version of the resource (3.0) was published 2014-05-23, probably using IPT v2.1. Since then, the IPT has upgraded the library it uses to serialise metadata to XML and back again, which may have fixed this old bug. Looking through old issues, I can't find it reported before though.

**Comment by silenius** [comment]
> When you re-edited the resource, did you remove all the offending escape characters?
I think so... to be honnest I'm not sure how those offending escape characters appeared as I'm not the one who filled the resource infos, but I'll let you know if the problem recurs.
